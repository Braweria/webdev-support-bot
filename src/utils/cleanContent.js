const linebreakPattern = /\n/gim;

module.exports = (msg) =>
  msg.cleanContent.replace(linebreakPattern, ' ').toLowerCase();
