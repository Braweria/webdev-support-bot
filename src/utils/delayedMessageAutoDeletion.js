const { missingRightsDeletion } = require('./errors');

const THIRTY_SECONDS_IN_MS = 30 * 1000;

/**
 *
 * @param {import('discord.js').Message} msg
 * @param {number} timeout
 */
const delayedMessageAutoDeletion = (msg, timeout = THIRTY_SECONDS_IN_MS) => {
  setTimeout(() => {
    msg.delete().catch(error => {
      console.warn("Couldn't delete message", error);
      msg.edit(missingRightsDeletion).catch(() => {
        console.info(
          "Couldn't edit message after trying to delete, probably removed by someone else.",
        );
      });
    });
  }, timeout);
};

module.exports = delayedMessageAutoDeletion;
