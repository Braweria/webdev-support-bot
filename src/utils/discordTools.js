const { providers } = require('./urlTools');
// eslint-disable-next-line no-unused-vars
const { Message } = require('discord.js');

/**
 *
 * @param {string} title
 * @param {string} url
 */
const createMarkdownLink = (title, url) =>
  `[${title}](${url.replace(/\)/g, '\\)')})`;

const BASE_DESCRIPTION = `
:bulb: *react with a number (:one:, :two:, ...) to filter your result*
:neutral_face: *react with ❌ or ✖️ to delete*
:point_up: *supports !npm, !composer, !mdn and !caniuse*
:gear: *issues? feature requests? head over to ${createMarkdownLink(
  'github',
  process.env.REPO_LINK,
)}*`;

/**
 *
 * @param {string} provider
 * @param {string} searchTerm
 * @param {string} url
 * @param {string} footerText
 * @param {string} description
 */
const createListEmbed = ({
  provider,
  searchTerm,
  url,
  footerText,
  description,
}) => {
  if (providers[provider]) {
    const { createTitle } = providers[provider];

    return createEmbed({
      provider,
      title: createTitle(searchTerm),
      url: url.substr(0, 2048),
      footerText,
      description,
    });
  }

  throw new Error('provider not implemented');
};

/**
 *
 * @param {string} provider
 * @param {string} title
 * @param {string} url
 * @param {string} footerText
 * @param {string} description
 * @param {array} fields
 */
const createEmbed = ({
  provider,
  title,
  url,
  footerText,
  description,
  fields = [],
  author = null,
}) => {
  if (providers[provider]) {
    const { color, icon } = providers[provider];

    return {
      embed: {
        title,
        color,
        url,
        footer: {
          icon_url: icon,
          text: footerText,
        },
        description,
        fields,
        author,
      },
    };
  }

  throw new Error('provider not implemented');
};

/**
 *
 * @param {Message} msg
 * @param {number} timeout
 */
const delayedAutoDeleteMessage = (msg, timeout = 30 * 1000) => {
  setTimeout(() => {
    msg.delete();
  }, timeout);
};

const DESCRIPTION_LENGTH_LIMIT = 72;
const SEPARATOR_LENGTH = 3;

/**
 * Cuts off the description of a package name
 * based on the maximum of possible characters before
 * a linebreak occurs, keeping words intact.
 *
 * @param {number} position
 * @param {string} name
 * @param {string} description
 */
const adjustDescriptionLength = (position, name, description) => {
  const positionLength = position.toString().length + 2;
  const nameLength = name.length;
  const descriptionLength = description.length;

  const currentLength =
    positionLength + nameLength + SEPARATOR_LENGTH + descriptionLength;

  if (currentLength > DESCRIPTION_LENGTH_LIMIT) {
    const availableSpace =
      DESCRIPTION_LENGTH_LIMIT - positionLength - nameLength - SEPARATOR_LENGTH;

    let hasHitLimit = false;

    const shortenedDescription = description
      .split(' ')
      .reduce((carry, part) => {
        if (hasHitLimit || carry.length + part.length > availableSpace) {
          hasHitLimit = true;
          return carry;
        }

        if (carry.length === 0) {
          return part;
        }

        return [carry, part].join(' ');
      }, '');

    return shortenedDescription + '...';
  }

  return description;
};

/**
 *
 * @param {number} index
 * @param {string} content
 */
const createMarkdownListItem = (index, content) => `${index + 1}. ${content}`;

/**
 *
 * @param {array} items
 */
const createDescription = items => items.concat(BASE_DESCRIPTION).join('\n');

module.exports = {
  createMarkdownLink,
  createListEmbed,
  createEmbed,
  delayedAutoDeleteMessage,
  adjustDescriptionLength,
  createMarkdownListItem,
  createDescription,
};
