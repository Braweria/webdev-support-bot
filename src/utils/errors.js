module.exports = {
  invalidResponse:
    'sorry, your request could not be processed. Please try again at a later time.',
  noResults: search =>
    `sorry, could not find anything for \`${search}\`. Autodeleting this response in 30 seconds.`,
  unknownError:
    'sorry, something went wrong. If this issue persists, please file an issue at https://github.com/ljosberinn/discord-mdn-bot',
  missingRightsDeletion: 'insufficient permissions: unable to delete message',
};
