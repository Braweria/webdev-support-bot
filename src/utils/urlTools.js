const { noResults, invalidResponse } = require('./errors');
const useData = require('./useData');
const delayedMessageAutoDeletion = require('./delayedMessageAutoDeletion');

const SEARCH_TERM = '%SEARCH%';
const TERM = '%TERM%';

const providers = {
  mdn: {
    search: `https://developer.mozilla.org/en-US/search?q=${SEARCH_TERM}`,
    direct: `https://developer.mozilla.org${TERM}`,
    color: 0x83d0f2,
    createTitle: (searchTerm) => `MDN results for *${searchTerm}*`,
    icon: 'https://avatars0.githubusercontent.com/u/7565578',
    help: '!mdn localStorage',
  },
  npm: {
    search: `https://www.npmjs.com/search/suggestions?q=${SEARCH_TERM}`,
    color: 0xfb3e44,
    createTitle: (searchTerm) => `NPM results for *${searchTerm}*`,
    icon: 'https://avatars0.githubusercontent.com/u/6078720',
    help: '!npm react',
  },
  composer: {
    search: `https://packagist.org/search.json?q=${SEARCH_TERM}`,
    direct: `https://packagist.org/packages/${TERM}`,
    color: 0xf28d1a,
    createTitle: (searchTerm) => `Packagist results for ${searchTerm}`,
    icon: 'https://packagist.org/bundles/packagistweb/img/logo-small.png',
    getExtendedInfoUrl: (package) =>
      `https://packagist.org/packages/${package}.json`,
    help: '!composer sentry',
  },
  caniuse: {
    search: `https://caniuse.com/process/query.php?search=${SEARCH_TERM}`,
    direct: `https://caniuse.com/#feat=${TERM}`,
    color: 0xdb5600,
    createTitle: (searchTerm) => `CanIUse results for *${searchTerm}*`,
    icon: 'https://caniuse.com/img/favicon-128.png',
    getExtendedInfoUrl: (text) =>
      `https://caniuse.com/process/get_feat_data.php?type=support-data&feat=${text}`,
    help: '!caniuse IntersectionObserver',
  },
  github: {
    search: `https://api.github.com/search/repositories?q=${SEARCH_TERM}`,
    direct: `https://github.com/${TERM}`,
    color: 0x24292e,
    createTitle: (searchTerm) => `GitHub results for *${searchTerm}*`,
    icon:
      'https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png',
    help: '!github react',
  },
  bundlephobia: {
    search: `https://api.npms.io/v2/search/suggestions?q=${SEARCH_TERM}`,
    direct: `https://bundlephobia.com/result?p=${TERM}`,
    color: 0xffffff,
    createTitle: (searchTerm) => `Bundlephobia results for *${searchTerm}*`,
    getExtendedInfoUrl: (package) =>
      `https://bundlephobia.com/api/size?package=${package}&record=true`,
    icon: 'https://bundlephobia.com/android-chrome-192x192.png',
    help: '!bundlephobia @chakra-ui/core',
  },
  spam: {
    color: 0xffffff,
    icon:
      'https://github.com/ljosberinn/webdev-support-bot/blob/master/logo.png?raw=true',
  },
};

const HELP_KEYWORD = '--help';

/**
 * dynamic regExp matching all possible Object.keys(providers) as keyword
 */
const KEYWORD_REGEXP = new RegExp(
  '^!(%PLACEHOLDER%)\\s+'.replace(
    '%PLACEHOLDER%',
    Object.keys(providers)
      .reduce((carry, keyword) => [...carry, keyword], [])
      .join('|'),
  ),
  'i',
);

/**
 *
 * @param {'mdn' | 'caniuse' | 'composer' | 'npm' | 'github' | 'bundlephobia'} provider
 * @param {string} search
 *
 * @returns {string}
 */
const getSearchUrl = (provider, search) => {
  if (providers[provider]) {
    return providers[provider].search.replace(SEARCH_TERM, encodeURI(search));
  }

  throw new Error(`provider not implemeted: ${provider}`);
};

/**
 * @param {'mdn' | 'caniuse' | 'composer' | 'npm' | 'github' | 'bundlephobia'} provider
 * @param {string} href
 */
const buildDirectUrl = (provider, href) => {
  if (providers[provider]) {
    return providers[provider].direct.replace(TERM, href);
  }

  throw new Error(`provider not implemeted: ${provider}`);
};

/**
 *
 * @param {'mdn' | 'caniuse' | 'composer' | 'npm' | 'github' | 'bundlephobia'} provider
 * @param {string} term
 *
 * @returns {string}
 */
const getExtendedInfoUrl = (provider, term) => {
  if (providers[provider] && providers[provider].getExtendedInfoUrl) {
    return providers[provider].getExtendedInfoUrl(term);
  }

  throw new Error(
    `provider or provider.getExtendedInfoUrl not implemented at provider: ${provider}`,
  );
};

/**
 *
 * @param {{
 *   msg: import('discord.js').Message,
 *   provider: 'caniuse' | 'composer' | 'mdn' | 'npm' | 'github' | 'bundlephobia',
 *   searchTerm: string,
 *   invalidData: (str: string) => boolean,
 *   sanitizeData?: (data: any) => any,
 *   headers?: Headers
 * }}
 *
 * @returns Promise<array|object|string>
 */
const getData = async ({
  msg,
  provider,
  searchTerm,
  sanitizeData,
  isInvalidData,
  headers,
}) => {
  const searchUrl = getSearchUrl(provider, searchTerm);
  const { error, json: data } = await useData(searchUrl, 'json', {
    headers,
  });

  if (error) {
    await msg.reply(invalidResponse);
    return;
  }

  const sanitizedData = sanitizeData ? sanitizeData(data) : data;

  if (isInvalidData(sanitizedData)) {
    const sentMessage = await msg.reply(noResults(searchTerm));

    delayedMessageAutoDeletion(sentMessage);
    return;
  }

  return sanitizedData;
};

module.exports = {
  providers,
  getSearchUrl,
  KEYWORD_REGEXP,
  buildDirectUrl,
  HELP_KEYWORD,
  getExtendedInfoUrl,
  getData,
};
