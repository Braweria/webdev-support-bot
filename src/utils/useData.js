const fetch = require('node-fetch');

const useData = async (url, type = 'json', headers = {}) => {
  const response = await fetch(url, headers);

  if (!response.ok) {
    return {
      error: true,
      json: null,
      text: null,
    };
  }

  if (type === 'json') {
    const json = await response.json();

    return {
      error: false,
      text: null,
      json,
    };
  }

  if (type === 'text') {
    const text = await response.text();

    return {
      error: false,
      json: null,
      text,
    };
  }
};

module.exports = useData;
