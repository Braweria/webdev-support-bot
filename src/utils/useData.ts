import fetch, {
  HeaderInit,
  RequestInfo,
  RequestInit,
  Response,
} from 'node-fetch';
import * as NodeCache from 'node-cache';

import {
  API_CACHE_ENTRIES_LIMIT,
  API_CACHE_EXPIRATION_IN_SECONDS,
  API_CACHE_REVALIDATION_WINDOW_IN_SECONDS,
} from '../env';

const apiCache = new NodeCache({
  checkperiod: parseInt(API_CACHE_REVALIDATION_WINDOW_IN_SECONDS, 10),
  stdTTL: parseInt(API_CACHE_EXPIRATION_IN_SECONDS, 10),
  maxKeys: parseInt(API_CACHE_ENTRIES_LIMIT, 10),
});

type ResponseTypes = 'json' | 'text';

type UnknownData<T> =
  | {
      error: true;
      json: null;
      text: null;
    }
  | {
      error: false;
      text: null;
      json: T;
    }
  | {
      error: false;
      json: null;
      text: string;
    };

type ResponseMapper<Data> = (response: Response) => Promise<Data>;
type FetchWithFormat<Format> = (
  url: RequestInfo,
  init?: RequestInit
) => Promise<Format>;

/**
 * Given a cache key and a function that maps the fetch response to a arbitrary
 * data structure, return a function when invoked, lower cases the cache key and
 * returns the cached formatted response when found.
 *
 * If the cache entry is _not_ found, delegate the call to an actual fetch implementation
 * to make the HTTP request. This response is then formatted with the function
 * above and saved into the cache. HTTP requests times/durations are also logged.
 */
const doFetch: <TParsedResponse>(
  cacheKey: string,
  mapper: ResponseMapper<TParsedResponse>
) => FetchWithFormat<TParsedResponse> = <TParsedResponse>(cacheKey, mapper) => {
  const casedCacheKey = cacheKey.toLowerCase();
  const cachedResponse = apiCache.get<TParsedResponse>(casedCacheKey);
  if (cachedResponse) {
    return async () => cachedResponse;
  }

  return async (url, fetchOptions) => {
    const timeLabel = `Time took for url=${encodeURIComponent(url.toString())}`;
    console.time(timeLabel);
    const response = await fetch(url, fetchOptions);
    console.timeEnd(timeLabel);
    const formattedResponse = await mapper(response);
    apiCache.set(casedCacheKey, formattedResponse);
    return formattedResponse;
  };
};

const responseMapper: <T>(
  type: ResponseTypes
) => (response: Response) => Promise<UnknownData<T>> = (
  type: string
) => async response => {
  if (!response.ok) {
    return {
      error: true,
      json: null,
      text: null,
    };
  }

  if (type === 'json') {
    const json = await response.json();
    return {
      error: false,
      json,
      text: null,
    };
  }

  if (type === 'text') {
    const text = await response.text();

    return {
      error: false,
      json: null,
      text,
    };
  }
};

export default async <T>(
  url: string,
  type: ResponseTypes = 'json',
  headers: HeaderInit = {}
): Promise<UnknownData<T>> => {
  return doFetch(url, responseMapper<T>(type))(url, { headers });
};
