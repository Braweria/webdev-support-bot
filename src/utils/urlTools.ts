import { noResults, invalidResponse } from './errors';
import useData from './useData';
import delayedMessageAutoDeletion from './delayedMessageAutoDeletion';
import { Provider } from './discordTools';
import { Message } from 'discord.js';
import { HeadersInit } from 'node-fetch';

const SEARCH_TERM = '%SEARCH%';
const TERM = '%TERM%';

type ProviderMap = {
  [key in Provider]: {
    search: string;
    color: number;
    createTitle: (term: string) => string;
    icon: string;
    help: string;
    direct?: string;
    getExtendedInfoUrl?: (pkg: string) => string;
  };
};

export const providers: ProviderMap = {
  mdn: {
    search: `https://developer.mozilla.org/en-US/search?q=${SEARCH_TERM}`,
    direct: `https://developer.mozilla.org${TERM}`,
    color: 0x83d0f2,
    createTitle: (searchTerm: string) => `MDN results for *${searchTerm}*`,
    icon: 'https://avatars0.githubusercontent.com/u/7565578',
    help: '!mdn localStorage',
  },
  npm: {
    search: `https://www.npmjs.com/search/suggestions?q=${SEARCH_TERM}`,
    color: 0xfb3e44,
    createTitle: (searchTerm: string) => `NPM results for *${searchTerm}*`,
    icon: 'https://avatars0.githubusercontent.com/u/6078720',
    help: '!npm react',
  },
  composer: {
    search: `https://packagist.org/search.json?q=${SEARCH_TERM}`,
    direct: `https://packagist.org/packages/${TERM}`,
    color: 0xf28d1a,
    createTitle: (searchTerm: string) => `Packagist results for ${searchTerm}`,
    icon: 'https://packagist.org/bundles/packagistweb/img/logo-small.png',
    getExtendedInfoUrl: (pkg: string) =>
      `https://packagist.org/packages/${pkg}.json`,
    help: '!composer sentry',
  },
  caniuse: {
    search: `https://caniuse.com/process/query.php?search=${SEARCH_TERM}`,
    direct: `https://caniuse.com/#feat=${TERM}`,
    color: 0xdb5600,
    createTitle: (searchTerm: string) => `CanIUse results for *${searchTerm}*`,
    icon: 'https://caniuse.com/img/favicon-128.png',
    getExtendedInfoUrl: (text: string) =>
      `https://caniuse.com/process/get_feat_data.php?type=support-data&feat=${text}`,
    help: '!caniuse IntersectionObserver',
  },
  github: {
    search: `https://api.github.com/search/repositories?q=${SEARCH_TERM}`,
    direct: `https://github.com/${TERM}`,
    color: 0x24292e,
    createTitle: (searchTerm: string) => `GitHub results for *${searchTerm}*`,
    icon:
      'https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png',
    help: '!github react',
  },
  bundlephobia: {
    search: `https://api.npms.io/v2/search/suggestions?q=${SEARCH_TERM}`,
    direct: `https://bundlephobia.com/result?p=${TERM}`,
    color: 0xffffff,
    createTitle: (searchTerm: string) =>
      `Bundlephobia results for *${searchTerm}*`,
    getExtendedInfoUrl: (pkg: string) =>
      `https://bundlephobia.com/api/size?package=${pkg}&record=true`,
    icon: 'https://bundlephobia.com/android-chrome-192x192.png',
    help: '!bundlephobia @chakra-ui/core',
  },
};

export const HELP_KEYWORD = '--help';
export const FORMATTING_KEYWORD = '!formatting';
export const CODE_KEYWORD = '!code';
export const VSCODE_KEYWORD = '!vscode';

/**
 * dynamic regExp matching all possible Object.keys(providers) as keyword
 */
export const KEYWORD_REGEXP = new RegExp(
  '^!(%PLACEHOLDER%)\\s+'.replace(
    '%PLACEHOLDER%',
    Object.keys(providers)
      .reduce((carry, keyword) => [...carry, keyword], [])
      .join('|')
  ),
  'i'
);

export const getSearchUrl = (provider: Provider, search: string) => {
  if (providers[provider]) {
    return providers[provider].search.replace(SEARCH_TERM, encodeURI(search));
  }

  throw new Error(`provider not implemeted: ${provider}`);
};

export const buildDirectUrl = (provider: Provider, href: string) => {
  if (providers[provider]) {
    return providers[provider].direct.replace(TERM, href);
  }

  throw new Error(`provider not implemeted: ${provider}`);
};

export const getExtendedInfoUrl = (provider: Provider, term: string) => {
  if (providers[provider] && providers[provider].getExtendedInfoUrl) {
    return providers[provider].getExtendedInfoUrl(term);
  }

  throw new Error(
    `provider or provider.getExtendedInfoUrl not implemented at provider: ${provider}`
  );
};

interface GetDataParams {
  msg: Message;
  provider: Provider;
  searchTerm: string;
  sanitizeData?: (data: any) => Partial<any>;
  isInvalidData: (data: any) => boolean;
  headers?: HeadersInit;
}

export const getData = async <T>({
  msg,
  provider,
  searchTerm,
  sanitizeData,
  isInvalidData,
  headers,
}: GetDataParams): Promise<Partial<T>> => {
  const searchUrl = getSearchUrl(provider, searchTerm);
  const { error, json: data } = await useData<T>(searchUrl, 'json', headers);

  if (error) {
    await msg.reply(invalidResponse);
    return;
  }

  const sanitizedData = sanitizeData ? sanitizeData(data) : data;

  if (isInvalidData(sanitizedData)) {
    const sentMessage = await msg.reply(noResults(searchTerm));

    delayedMessageAutoDeletion(sentMessage);
    return;
  }

  return sanitizedData;
};
