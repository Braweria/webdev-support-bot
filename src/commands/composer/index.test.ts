import { buildComposerQueryHandler } from './index';
import { getData } from '../../utils/urlTools';
import { PackagistResponse } from './types';
import useData from '../../utils/useData';
import { getChosenResult } from '../../utils/discordTools';
import * as errors from '../../utils/errors';

describe('handleComposerQuery', () => {
  const sendMock = jest.fn();
  const replyMock = jest.fn();
  const msg: any = {
    channel: { send: sendMock },
    reply: replyMock,
  };

  const fetch: jest.MockedFunction<typeof getData> = jest.fn();
  const fetchUse: jest.MockedFunction<typeof useData> = jest.fn();
  const choose: jest.MockedFunction<typeof getChosenResult> = jest.fn();

  afterEach(() => jest.resetAllMocks());

  test('does nothing when packagist request is invalid', async () => {
    fetch.mockResolvedValue(null);
    const handler = buildComposerQueryHandler(fetch);
    await handler(msg, 'search term');

    expect(fetch).toBeCalledWith({
      isInvalidData: expect.any(Function),
      msg,
      provider: 'composer',
      searchTerm: 'search term',
    });

    expect(msg.channel.send).not.toHaveBeenCalled();
  });

  test('returns when chosen result is not given', async () => {
    const packageRes: PackagistResponse = {
      results: [
        {
          name: 'Symfony',
          description: 'Better than Laravel',
          url: 'https://symfony.com',
          repository: 'https://github.com/symfony/symfony',
          downloads: 10000,
          favers: 10000,
        },
      ],
      total: 1,
    };

    fetch.mockResolvedValue(packageRes);
    choose.mockResolvedValue(undefined);
    msg.channel.send.mockResolvedValue(null);

    const handler = buildComposerQueryHandler(fetch, fetchUse, choose);
    await handler(msg, 'search term');

    expect(fetch).toBeCalledWith({
      isInvalidData: expect.any(Function),
      msg,
      provider: 'composer',
      searchTerm: 'search term',
    });

    expect(msg.channel.send).toHaveBeenCalledWith(
      expect.objectContaining({
        embed: expect.objectContaining({
          description: expect.stringContaining('Symfony'),
        }),
      })
    );
    expect(fetchUse).not.toBeCalled();
  });

  test('calls the use useData method and fails', async () => {
    const packageRes: PackagistResponse = {
      results: [
        {
          name: 'Symfony',
          description: 'Better than Laravel',
          url: 'https://symfony.com',
          repository: 'https://github.com/symfony/symfony',
          downloads: 10000,
          favers: 10000,
        },
      ],
      total: 1,
    };

    fetch.mockResolvedValue(packageRes);
    choose.mockResolvedValue({
      name: 'Symfony',
      description: 'Better than Laravel',
      url: 'https://symfony.com',
      repository: 'https://github.com/symfony/symfony',
      downloads: 10000,
      stars: 10000,
    });
    fetchUse.mockResolvedValue({
      error: true,
      json: null,
      text: null,
    });

    msg.channel.send.mockResolvedValue(null);

    const handler = buildComposerQueryHandler(fetch, fetchUse, choose);
    await handler(msg, 'search term');

    expect(fetch).toBeCalledWith({
      isInvalidData: expect.any(Function),
      msg,
      provider: 'composer',
      searchTerm: 'search term',
    });

    const payload = msg.channel.send.mock.calls[0][0];
    expect(payload).toMatchSnapshot();
    expect(msg.reply).toBeCalledWith(errors.invalidResponse);
  });

  test('calls the use useData method and succeeds', async () => {
    const editMock = jest.fn();
    sendMock.mockResolvedValue({ edit: editMock });

    const packageRes: PackagistResponse = {
      results: [
        {
          name: 'Symfony',
          description: 'Better than Laravel',
          url: 'https://symfony.com',
          repository: 'https://github.com/symfony/symfony',
          downloads: 10000,
          favers: 10000,
        },
      ],
      total: 1,
    };

    fetch.mockResolvedValue(packageRes);
    choose.mockResolvedValue({
      name: 'Symfony',
      description: 'Better than Laravel',
      url: 'https://symfony.com',
      repository: 'https://github.com/symfony/symfony',
      downloads: 10000,
      stars: 10000,
    });
    fetchUse.mockResolvedValue({
      error: false,
      json: {
        package: {
          name: 'Symfony',
          description: 'Better than Laravel',
          time: new Date('2020-01-01'),
          maintainers: [
            {
              avatar_url: 'http://avatar',
              name: 'A Name',
            },
          ],
          versions: {
            'dev-master': {
              name: 'master',
              description: 'master',
              keywords: ['PHP'],
              homepage: 'https://github.com/symfony/symfony',
              version: '1.0.0',
              version_normalized: '1.0.0',
              license: ['MIT'],
              time: new Date('2020-01-01'),
              require: ['php', '7'],
              authors: [{ name: 'Example ' }],
            },
            '1.0.0': {
              name: 'master',
              description: 'master',
              keywords: ['PHP'],
              homepage: 'https://github.com/symfony/symfony',
              version: '1.0.0',
              version_normalized: '1.0.0',
              license: ['MIT'],
              time: new Date('2020-01-01'),
              require: ['php', '7'],
              authors: [{ name: 'Example ' }],
            },
          },
          type: 'Framework',
          repository: 'https://github.com/symfony/symfony',
          github_stars: 10000,
          github_watchers: 10000,
          github_forks: 1000,
          github_open_issues: 0,
          language: 'PHP',
          dependents: 10,
          suggesters: 1000,
          downloads: {
            total: 1000000,
            monthly: 1234,
            daily: 1234,
          },
          favers: 1234,
        },
      },
      text: null,
    });

    msg.channel.send.mockResolvedValue(null);

    const handler = buildComposerQueryHandler(fetch, fetchUse, choose);
    await handler(msg, 'search term');

    expect(fetch).toBeCalledWith({
      isInvalidData: expect.any(Function),
      msg,
      provider: 'composer',
      searchTerm: 'search term',
    });

    const payload = msg.channel.send.mock.calls[0][0];
    expect(payload).toMatchSnapshot();

    expect(msg.channel.send).toHaveBeenCalledWith(
      expect.objectContaining({
        embed: expect.objectContaining({
          description: expect.stringContaining('Symfony'),
        }),
      })
    );

    const editPayload = sendMock.mock.calls[0][0];
    expect(editPayload).toMatchSnapshot();
  });
});
