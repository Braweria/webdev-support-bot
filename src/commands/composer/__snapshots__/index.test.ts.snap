// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleComposerQuery calls the use useData method and fails 1`] = `
Object {
  "embed": Object {
    "author": null,
    "color": 15895834,
    "description": "1. [**Symfony** - *Better than Laravel*](https://symfony.com)

:bulb: *react with a number (:one:, :two:, ...) to filter your result*
:neutral_face: *react with \`❌\` to delete*
:point_up: *supports \`!mdn\`, \`!github\`, \`!caniuse\`, \`!npm\`, \`!composer\`, \`!bundlephobia\`, and \`!php\`*
:gear: *issues? feature requests? head over to [github](https://github.com/ljosberinn/webdev-support-bot)*",
    "fields": Array [],
    "footer": Object {
      "iconURL": "https://packagist.org/bundles/packagistweb/img/logo-small.png",
      "text": "1 packages found",
    },
    "title": "Packagist results for search term",
    "url": "https://packagist.org/?query=search%20term",
  },
}
`;

exports[`handleComposerQuery calls the use useData method and succeeds 1`] = `
Object {
  "embed": Object {
    "author": null,
    "color": 15895834,
    "description": "1. [**Symfony** - *Better than Laravel*](https://symfony.com)

:bulb: *react with a number (:one:, :two:, ...) to filter your result*
:neutral_face: *react with \`❌\` to delete*
:point_up: *supports \`!mdn\`, \`!github\`, \`!caniuse\`, \`!npm\`, \`!composer\`, \`!bundlephobia\`, and \`!php\`*
:gear: *issues? feature requests? head over to [github](https://github.com/ljosberinn/webdev-support-bot)*",
    "fields": Array [],
    "footer": Object {
      "iconURL": "https://packagist.org/bundles/packagistweb/img/logo-small.png",
      "text": "1 packages found",
    },
    "title": "Packagist results for search term",
    "url": "https://packagist.org/?query=search%20term",
  },
}
`;

exports[`handleComposerQuery calls the use useData method and succeeds 2`] = `
Object {
  "embed": Object {
    "author": null,
    "color": 15895834,
    "description": "1. [**Symfony** - *Better than Laravel*](https://symfony.com)

:bulb: *react with a number (:one:, :two:, ...) to filter your result*
:neutral_face: *react with \`❌\` to delete*
:point_up: *supports \`!mdn\`, \`!github\`, \`!caniuse\`, \`!npm\`, \`!composer\`, \`!bundlephobia\`, and \`!php\`*
:gear: *issues? feature requests? head over to [github](https://github.com/ljosberinn/webdev-support-bot)*",
    "fields": Array [],
    "footer": Object {
      "iconURL": "https://packagist.org/bundles/packagistweb/img/logo-small.png",
      "text": "1 packages found",
    },
    "title": "Packagist results for search term",
    "url": "https://packagist.org/?query=search%20term",
  },
}
`;
